vumi_message_store/memory_backend_manager.py

```diff
@@ -0,0 +1,371 @@
+        wait for the deferred to fire.
+        """
+        if self._is_sync:
+            return func(*args, **kw)
+
+        # Add some latency to catch things that don't wait on deferreds. We
+        # can't use deferLater() here because we want to keep track of the
+        # delayed call object.
+        d = Deferred()
+        delayed = reactor.callLater(
+            0.002, lambda _: d.callback(func(*args, **kw)), None)
+        self._delayed_calls.append(delayed)
+        return d
+
+    def _get_bucket(self, bucket_name):
+        """
+        Get the data for a bucket.
+        """
+        return self._buckets.setdefault(bucket_name, {
+            "objects": {},
+            "indexes": {},
+        })
+
+    def store_object(self, fake_object):
+        """
+        Store an object.
+        """
+        bucket = self._get_bucket(fake_object._bucket_name)
+        bucket["objects"][fake_object.key] = {
+            "content_type": fake_object._current_data["content_type"],
+            "data": json.dumps(fake_object._current_data["data"]),
+        }
+        indexes = {}
+        for index_name, index_value in fake_object._current_data["indexes"]:
+            indexes.setdefault(index_name, []).append(
+                (index_value, fake_object.key))
+        for index_name, index_values in indexes.iteritems():
+            old_indexes = bucket["indexes"].get(index_name, [])
+            new_indexes = [
+                (value, key) for value, key in old_indexes
+                if key != fake_object.key]
+            new_indexes.extend(index_values)
+            new_indexes.sort()
+            bucket["indexes"][index_name] = new_indexes
+
+    def _reload_object_state(self, fake_object):
+        """
+        Set an object's data from stored state.
+        """
+        bucket = self._get_bucket(fake_object._bucket_name)
+        obj_state = bucket["objects"].get(fake_object.key)
+        if obj_state is None:
+            fake_object._current_data["data"] = None
+            return
+        fake_object._current_data["data"] = json.loads(obj_state["data"])
+        fake_object._current_data["content_type"] = obj_state["content_type"]
+        fake_object._current_data["indexes"] = set()
+        for index_name, index_data in bucket["indexes"].iteritems():
+            for index_value, key in index_data:
+                if key == fake_object.key:
+                    fake_object._current_data["indexes"].add(
+                        (index_name, index_value))
+
+    def get_index_page(self, bucket_name, index_name, start_value, end_value,
+                       return_terms=False, max_results=None,
+                       continuation=None):
+        """
+        Perform an index query.
+        """
+        # Get all matching results
+        index = self._get_bucket(bucket_name)["indexes"].get(index_name, [])
+        if end_value is None:
+            in_range = lambda v: v[0] == start_value
```

Comment:
If we don't have an `end_value`, we're doing an exact match query instead of a range query.